core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/PathTrackingInputFormat.java

```diff
@@ -99,14 +91,14 @@ public static Schema getTextOutputSchema(@Nullable String pathField) {
     if ("avro".equalsIgnoreCase(format)) {
       RecordReader<AvroKey<GenericRecord>, NullWritable> delegate = (new AvroKeyInputFormat<GenericRecord>())
         .createRecordReader(split, context);
-      return new TrackingAvroRecordReader(delegate, pathField, path, schema, job);
+      return new TrackingAvroRecordReader(delegate, pathField, path, Schema.parseJson(schema));
     } else if ("parquet".equalsIgnoreCase(format)) {
       RecordReader<Void, GenericRecord> delegate = (new AvroParquetInputFormat<GenericRecord>())
         .createRecordReader(split, context);
-      return new TrackingParquetRecordReader(delegate, pathField, path, schema, job);
+      return new TrackingParquetRecordReader(delegate, pathField, path, Schema.parseJson(schema));
     } else {
       RecordReader<LongWritable, Text> delegate = (new TextInputFormat()).createRecordReader(split, context);
-      return new TrackingTextRecordReader(delegate, pathField, path);
+      return new TrackingTextRecordReader(delegate, pathField, path, getTextOutputSchema(pathField));
```

Comment:
Do Schema.parseJson() before this if else section and pass the schema to each record reader. As mentioned before, you shouldn't be using getTextOutputSchema() here, as it makes it easier to accidentally break the code as it evolves.